[WIP] Put database operations in separate package

vbmade2000 opened the pull request:
Fixes #2

vbmade2000 committed:
Put database operations in separate package

Fixes #2

sbinet commented on database/boltdb.go:
```diff
@@ -0,0 +1,135 @@
+import (
+	"bytes"
+	"errors"
+	bolt "github.com/coreos/bbolt"
```
please use `goimports` to format your code.

https://golang.org/x/tools/cmd/goimports

vbmade2000 replied:
Sure. Actually entire formatting is yet to be done.

sbinet commented on database/boltdb.go:
```diff
@@ -0,0 +1,135 @@
+package database
```
we are missing a copyright header here.

vbmade2000 replied:
I'll add that.

sbinet commented on database/boltdb.go:
```diff
@@ -0,0 +1,135 @@
+type BoltDataStore struct {
+	Db *bolt.DB
+}
+type byteSlice [][]byte
```
I would move this closer to the methods defined on that type :)

vbmade2000 replied:
Sure.

sbinet commented on database/common/common.go:
```diff
@@ -0,0 +1,9 @@
+package common
```
we are missing a copyright header here as well.

vbmade2000 replied:
I'll add that.

sbinet commented on database/common/common.go:
```diff
@@ -0,0 +1,9 @@
+   Usubscribe(user, list string) error
+   Lists() ([]string, error)
+   Users() ([]string, error)
+}
```
we are missing a `\n` there.

vbmade2000 replied:
Yes. I'll check that too.

sbinet commented on server.go:
```diff
@@ -31,10 +32,11 @@ var (
+	db        *bolt.DB
+	sck       net.Listener
+	msg       chan *Message
+	boltStore *database.BoltDataStore
```
using the explicit type here kind of defeats the purpose of introducing an interface :)

we should drop the `db *bolt.DB` field and replace `boltStore` with:

```go
type Server struct {
    // ...
    db database.Store
}
```

assuming the interface `Store` is defined in the `database` package:

```go
package database

type Store interface {
     Subscribers(list string) ([]string, error)
￼     Subscribe(user, list string) error
     Usubscribe(user, list string) error
     Lists() ([]string, error)
     Users() ([]string, error)
}
```

then, the config should be modified to specify the database name/type to use.
name to be used later on with your `database.New()` or, rather: `database.Open("boltdb", cfg.Database)`

vbmade2000 replied:
Agreed with that. Conrete type is to be replaced by interface but BoltDBStore didn't implement all the methods of interface so it was remaining. I'll do that now.

vbmade2000 commented:
I think We should also add isSubscribed() to Store interface and move it to boltdb.go. What do you think ?

sbinet commented:
this can be a helper function implemented in terms of the `Store` interface:

```go
func isSubscribed(db Store, user, list string) bool {
    users, err := db.Subscribers(list)
    if err != nil { return false }
    for _, u := range users {
        if u == user { return true }
    }
    return false
}
```

no need to add it to the interface.

vbmade2000 commented:
I modified as following.
```
func (srv *Server) isSubscribed(db common.Store, user, list string) bool {
	var o bool
	usr := []byte(user)
	key := []byte(list)
	err := db.Update(func(tx *bolt.Tx) error {
		b := tx.Bucket(subBucket)
		v := b.Get(key)
		for _, v := range bytes.Split(v, []byte(",")) {
			if bytes.Equal(v, usr) {
				o = true
				return nil
			}
		}
		return nil
	})
	if err != nil {
		return false
	}
	return o
}
```

It raises `db.Update undefined.(type common.Store has no method or field Db)`

I can add GetDB() method to Store interface but not sure is it a proper way.

sbinet commented:
I must say I don't understand how you get that error.

also: as I said in https://github.com/sbinet-alt63/strew/pull/7#issuecomment-372648049 `isSubscribed` can be implemented in terms of `Store.Subscribers()`.
no need to reach into `boltdb.Update` and its buckets.

vbmade2000 commented:
It already has db.Update() call. I just added `db common.Store` in function arguments.

sbinet commented:
closed in favour of #10
